Close reasons

nateps opened the pull request:
Add support for differentiating between different types of WebSocket closures by passing `code` and `reason` values to `ws.close()`. This is helpful, because a client may wish to perform different reconnection logic depending on the reason for disconnection.

In particular, it is useful to be able to differentiate intentional disconnections where the server wishes to indicate to the client that it should not reconnect at all or the client should delay for a more extended time before it reconnects. If, for example, a client's application version is no longer supported, the server may wish to disconnect with a reason that indicates the client should not attempt reconnection and it should refresh or update its code first.

In addition, this PR includes a full shim of Node's `stream.destroy()`, borrowed from current master of Node. The minimal shim was not consistent with the current implementation of stream.destroy() in Node 8+, particularly because it didn't emit passed in errors as an 'error' event. If we're shimming this, I believe we should use the full implementation from current Node for more consistency in API.

Including a shim of `stream.destroy()` is a key part of this PR, since it now produces a different close code than `stream.end()`

nateps committed:
shim destroy with full implementation from master of Node.js

The minimal shim was not consistent with the current implementation of stream.destroy() in Node 8+, particularly because it didn't emit passed in errors as an 'error' event. If we're shimming this, I believe we should use the full implementation from current Node for more consistency in API.

nateps committed:
provide code and reason values to CloseEvents

Add support for differentiating between different types of WebSocket closures by passing `close` and `reason` values to `ws.close()`. This is helpful, because a client may wish to perform different reconneciton logic depending on the reason for disconnection.

In particular, it is useful to be able to differentiate intentional disconnections where the server wishes to indicate to the client that it should not reconnect at all or the client should delay for a more extended time before it reconnects. If, for example, a client's application version is no longer supported, the server may wish to disconnect with a reason that indicates the client should not attempt reconnection and it should refresh or update its code first.

nateps committed:
clarify close reason documentation in readme

nateps committed:
spelling fix in readme

nateps committed:
correct statement about 1005 in readme

nateps committed:
fix eslint issues

gkubisa commented on test/close-status.js:
```diff
@@ -0,0 +1,101 @@
+const CUSTOM_CODE = 4000
+
+module.exports = function() {
+    it('should emit CloseEvent with code 1000 and reason "stream end" on clientStream.end()', function (done) {
```
Great tests!

Could we also get the same set of tests where the `close` event listener is attached to `this.serverWebSocket`? It's just to make sure that the server also gets the code and reason as expected.

And for consistency, could we change `.onclose = ...` to `.on('close', ...)`?

nateps replied:
Good points. I updated to run all tests against both `this.clientWebSocket` and `this.serverWebSocket`.

In regard to the close listener, these actually are different. `ws` emits a Node.js-style `on('close', ...)` event that has different arguments than the standard event interface of WebSockets. The WebSocket `CloseEvent` is emitted if you use `addEventListener('close', ...)` or `onclose = `. I decided to add tests for both.

gkubisa commented on destroy.js:
```diff
@@ -0,0 +1,91 @@
+// Copied from:
```
As this code has been copied and pasted from node, I don't think we need test coverage on it. Please add `/* istanbul ignore file */` at the top of this file.

nateps replied:
Agreed. I see mention of `/* istanbul ignore file */`, but it didn't seem to work (adding did not change coverage appropriately). I wonder if it is actually supported?

I used `/* istanbul ignore next */` before each function instead.

gkubisa replied:
No problem, I excluded that file on the command line using the `-x` flag.

gkubisa reviewed (changes requested):
Looks great to me! Thanks @nateps :+1:

Could you just make the few little changes before I merge, please?

nateps committed:
add istanbul ignore statements to code copied from node.js

nateps committed:
add tests for both WebSocket standard CloseEvent & ws's Node.js-style 'close' event

WebSockets emit CloseEvent objects to the `onclose` method (if defined) and any listeners added with addEventListener. The `ws` module also emits a Node.js-style 'close' event with different arguments. Add tests for both signatures.

In addition, test that these events are emitted to both clientWebSocket and serverWebSocket.

nateps committed:
fix eslint issues

nateps commented:
Regarding coverage testing, Coveralls is not reporting the same output as Istanbul anymore. They made some changes in Coveralls 3.0, and branch coverage is computed within ignored statements.

See output from Travis build showing coverage (with ignores) is 100%: https://travis-ci.org/Teamwork/websocket-json-stream/jobs/408183093

Apparently this can be worked around by downgrading the version of coveralls or by turning off the branch coverage setting for the repo:
https://github.com/lemurheavy/coveralls-public/issues/939
https://github.com/nickmerwin/node-coveralls/issues/158#issuecomment-286318287

gkubisa reviewed (approved):
Everything is in order now. Thanks for the excellent contribution :boom:

nateps commented:
Awesome. Thanks for the review!

On Thu, Jul 26, 2018 at 4:59 AM, Greg Kubisa <notifications@github.com>
wrote:

> Merged #1 <https://github.com/Teamwork/websocket-json-stream/pull/1>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Teamwork/websocket-json-stream/pull/1#event-1754835341>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOLNCPhXMTob7cxyfHMjr1DCOBKIq5nks5uKa8pgaJpZM4VfcrQ>
> .
>
